File: examples/similarity.rs
use anyhow::Result;
use edgebert::{Model, ModelType, cosine_similarity};

fn main() -> Result<()> {
    let model = Model::from_pretrained(ModelType::MiniLML6V2)?;

    let texts = vec!["Hello world", "How are you?", "Goodbye world"];
    let embeddings = model.encode(texts.clone(), false)?; // normalize=false

    for (i, embedding) in embeddings.iter().enumerate() {
        let norm: f32 = embedding.iter().map(|x| x*x).sum::<f32>().sqrt();
        let n = embedding.len().min(10);
        println!("Text: {} | Norm: {:.4} | First 10 dims: {:?}", texts[i], norm, &embedding[0..n]);
    }

    for i in 0..embeddings.len() {
        for j in (i+1)..embeddings.len() {
            let sim = cosine_similarity(&embeddings[i], &embeddings[j]);
            println!("Cosine similarity ({} <-> {}) = {:.4}", texts[i], texts[j], sim);
        }
    }

    Ok(())
}

File: examples/similarity.py
from sentence_transformers import SentenceTransformer, util
import numpy as np

# Load the same MiniLM model (MiniLM-L6-v2)
model = SentenceTransformer('all-MiniLM-L6-v2').to('cpu')

texts = ["Hello world", "How are you?", "Goodbye world"]

embeddings = model.encode(texts, convert_to_numpy=True, normalize_embeddings=True)

for i, emb in enumerate(embeddings):
    norm = np.linalg.norm(emb)
    print(f"Text: {texts[i]} | Norm: {norm:.4f} | First 10 dims: {emb[:10]}")

for i in range(len(texts)):
    for j in range(i+1, len(texts)):
        a, b = embeddings[i], embeddings[j]
        cos_sim = np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))
        print(f"Cosine similarity ({texts[i]} <-> {texts[j]}) = {cos_sim:.4f}")